deepwell/src/services/site_member/service.rs

```diff
@@ -0,0 +1,158 @@
+/*
+ * services/site_member/service.rs
+ *
+ * DEEPWELL - Wikijump API provider and database manager
+ * Copyright (C) 2019-2023 Wikijump Team
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+use super::prelude::*;
+use crate::models::site_member::{self, Entity as SiteMember, Model as SiteMemberModel};
+
+#[derive(Debug)]
+pub struct SiteMemberService;
+
+impl SiteMemberService {
+    /// Add a user to a site.
+    pub async fn add(
+        ctx: &ServiceContext<'_>,
+        SiteMembership { site_id, user_id }: SiteMembership,
+    ) -> Result<Option<SiteMemberModel>> {
+        let txn = ctx.transaction();
+        tide::log::info!(
+            "Adding membership of user with ID {user_id} to site ID {site_id}"
+        );
+
+        // If the user is already a member of the target site, discontinue.
+        if Self::get_optional(ctx, SiteMembership { site_id, user_id })
+            .await?
+            .is_some()
+        {
+            return Ok(None);
+        }
+
+        // TODO: Check for membership qualifications (e.g. bans).
+
+        // Insert new membership.
+        let model = site_member::ActiveModel {
+            user_id: Set(user_id),
+            site_id: Set(site_id),
+            deleted_at: Set(None),
```

Comment:
Remove this, unnecessary.

Reply:
Right, forgot that Option implements Default

Reply:
It's not `Default` (Rust trait), it's `DEFAULT` (database column setting).

All nullable columns have an implicit `DEFAULT NULL` basically.

Reply:
Well yeah but also that `Set(None)` would automatically be included in `..Default::default` is what I meant

Reply:
Actually no, the `Default` here is `Unset`, which is analogous to a SQL query where the value is simply excluded.

But yeah I get you.